openedx/core/djangoapps/safe_sessions/middleware.py

```diff
@@ -0,0 +1,453 @@
+        """
+        Return the user_id stored in the session of the request.
+        """
+        # Starting in django 1.8, the user_id is now serialized
+        # as a string in the session.  Before, it was stored
+        # directly as an integer. If back-porting to prior to
+        # django 1.8, replace the implementation of this method
+        # with:
+        # return request.session[SESSION_KEY]
+        from django.contrib.auth import _get_user_session_key
+        try:
+            return _get_user_session_key(request)
+        except KeyError:
+            return None
+
+    @staticmethod
+    def set_user_id_in_session(request, user):
+        """
+        Stores the user_id in the session of the request.
+        Used by unit tests.
+        """
+        # Starting in django 1.8, the user_id is now serialized
+        # as a string in the session.  Before, it was stored
+        # directly as an integer. If back-porting to prior to
+        # django 1.8, replace the implementation of this method
+        # with:
+        # request.session[SESSION_KEY] = user.id
+        request.session[SESSION_KEY] = user._meta.pk.value_to_string(user)  # pylint: disable=protected-access
+
+    @staticmethod
+    def update_with_safe_session_cookie(cookies, user_id):
+        """
+        Replaces the session_id in the session cookie with a freshly
+        computed safe_cookie_data.
+        """
+        # Create safe cookie data that binds the user with the session
+        # in place of just storing the session_key in the cookie.
+        safe_cookie_data = SafeCookieData.create(
+            cookies[settings.SESSION_COOKIE_NAME].value,
+            user_id,
+        )
+
+        # Update the cookie's value with the safe_cookie_data.
+        cookies[settings.SESSION_COOKIE_NAME] = unicode(safe_cookie_data)
+
+
+def _mark_cookie_for_deletion(request):
+    """
+    Updates the given request object to designate that the session
+    cookie should be deleted.
+    """
+    request.need_to_delete_cookie = True
+
+
+def _is_cookie_marked_for_deletion(request):
+    """
+    Returns whether the session cookie has been designated for deletion
+    in the given request object.
+    """
+    return getattr(request, 'need_to_delete_cookie', False)
+
+
+def _is_cookie_present(response):
+    """
+    Returns whether the session cookie is present in the response.
+    """
+    return (
+        response.cookies.get(settings.SESSION_COOKIE_NAME) and  # cookie in response
+        response.cookies[settings.SESSION_COOKIE_NAME].value  # cookie is not empty
+    )
+
+
+def _delete_cookie(response):
+    """
+    Delete the cookie by setting the expiration to a date in the past,
+    while maintaining the domain, secure, and httponly settings.
+    """
+    response.set_cookie(
+        settings.SESSION_COOKIE_NAME,
+        max_age=0,
+        expires='Thu, 01-Jan-1970 00:00:00 GMT',
+        domain=settings.SESSION_COOKIE_DOMAIN,
```

Comment:
Thanks for pointing this out @chrisndodge 
How does microsite override the django code then since django also creates cookies using the `settings`?  For example, see: https://github.com/django/django/blob/master/django/contrib/sessions/middleware.py#L51